### PR TITLE
[extension/jaegerremotesampling] Add the HTTP server implementation

### LINKED ISSUE #6632: Add the HTTP server

### PR DESCRIPTION
**Pull request by jpkrohling**
Closes #6632

This PR gives an initial implementation for the HTTP server providing the sampling strategies. The contents are currently coming from an initial simple implementation and will be changed in the next PRs to use the existing config manager proxy.

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

### CODE REVIEW

File: extension/jaegerremotesampling/extension.go
Comment (bogdandrutu) [inline extension/jaegerremotesampling/extension.go:59]: We ensure that "Start/Stop" is called in this order and only once. There is a valid use-case "Start/Stop/Start/Stop" but not "Start/Start".

File: extension/jaegerremotesampling/internal/http.go
Comment (bogdandrutu) [inline extension/jaegerremotesampling/internal/http.go:91]: Need to wait for `shutdownWG.Wait()` before returning.

Comment (jpkrohling) [reply extension/jaegerremotesampling/internal/http.go:91]: True, fixed.

File: extension/jaegerremotesampling/extension.go
Comment (jpkrohling) [reply extension/jaegerremotesampling/extension.go:59]: Alright. I added it after I myself made a mistake calling the API from the tests, but if there's a guarantee that no Start/Start, or Shutdown (without Start) will be called, then we indeed don't need to track whether the component has started.

**Comment by jpkrohling** [comment]
@open-telemetry/collector-contrib-approvers, could I get a review on this PR?

**Comment by jpkrohling** [comment]
@dmitryax, @djaglowski, could one of you please review this PR?

File: extension/jaegerremotesampling/README.md
Comment (bogdandrutu) [inline extension/jaegerremotesampling/README.md:25]: Looking at this config, kind of don't like, especially the "remote" vs "strategy_file". What about:

```yaml
  jaegerremotesampling:
    grpc:
      endpoint: :15251
    strategy:
      remote:
        endpoint: jaeger-collector:14250
  jaegerremotesampling/1:
    http:
      endpoint: :5878
    strategy:
       file: /etc/otel/sampling_strategies.json
```

So we separate the strategy config vs the other config.

File: extension/jaegerremotesampling/config.go
Comment (bogdandrutu) [inline extension/jaegerremotesampling/config.go:27]: Do we need to support http as well?

File: extension/jaegerremotesampling/README.md
Comment (jpkrohling) [reply extension/jaegerremotesampling/README.md:25]: Looks better, but I would change it to `source` or `strategies`.  Strategy is ambiguous here, as it could mean the "retrieval strategy" or "sampling strategies". `source` is clearer, so, I'll change it to that, but let me know if you have another suggestion.

File: extension/jaegerremotesampling/config.go
Comment (jpkrohling) [reply extension/jaegerremotesampling/config.go:27]: Yes, if we want to be compatible with Jaeger clients.

File: extension/jaegerremotesampling/README.md
Comment (jpkrohling) [reply extension/jaegerremotesampling/README.md:25]: PR updated.

**Comment by jpkrohling** [comment]
PR rebased.

**Comment by jpkrohling** [comment]
PR rebased. @bogdandrutu, can you take a look?

**Comment by jpkrohling** [comment]
@bogdandrutu or @open-telemetry/collector-contrib-approvers, could you take a look? Once I have an approval, I'll rebase and merge.

File: extension/jaegerremotesampling/README.md
Comment (bogdandrutu) [reply extension/jaegerremotesampling/README.md:25]: I don't see the update in the readme. Still using `strategy_file`

**Review by bogdandrutu — COMMENTED** [review]
LGTM the config as in code, readme needs update as well as some more documentation in the code.

File: extension/jaegerremotesampling/README.md
Comment (jpkrohling) [reply extension/jaegerremotesampling/README.md:25]: Readme fixed

File: extension/jaegerremotesampling/config.go
Comment (jpkrohling) [reply extension/jaegerremotesampling/config.go]: Done

**Comment by jpkrohling** [comment]
PR updated

**Comment by jpkrohling** [comment]
@bogdandrutu could you please review this one? [For some reason](https://github.com/open-telemetry/opentelemetry-collector/pull/4721#discussion_r789962357), we need to ping people to review 🙂

File: extension/jaegerremotesampling/config.go
Comment (bogdandrutu) [inline extension/jaegerremotesampling/config.go:47]: nit: can be non pointer and compare with empty string.

File: extension/jaegerremotesampling/config.go
Comment (bogdandrutu) [inline extension/jaegerremotesampling/config.go:35]: Do you allow enabling one of the http/grpc? How do you know when one is enabled vs the other? Probably you need this to be pointer.

File: extension/jaegerremotesampling/internal/clientconfigmanager.go
Comment (bogdandrutu) [inline extension/jaegerremotesampling/internal/clientconfigmanager.go:25]: Some comments about what this struct do?

File: extension/jaegerremotesampling/config.go
Comment (jpkrohling) [reply extension/jaegerremotesampling/config.go:35]: Both are enabled by default, but your question made me realize that users might not want to enable both, so, we need a way to either allow users to disable those ports, or only open when explicitly set. I think I prefer the second, so, making it a pointer will indeed be a good solution.

File: extension/jaegerremotesampling/config.go
Comment (jpkrohling) [reply extension/jaegerremotesampling/config.go:44]: I'm ambivalent on this. There have been cases of people who did have problems with gRPC in the past, but people should really prefer gRPC here. I'd leave it as is, and expand with "http_remote" in the future if there are demands for this. Keeping this simply as "remote" conveys the idea that this is the preferred/default transport.

File: extension/jaegerremotesampling/config.go
Comment (jpkrohling) [reply extension/jaegerremotesampling/config.go:50]: I think it reads better the way it is: "reload interval of the data from the source".

File: extension/jaegerremotesampling/internal/http.go
Comment (jpkrohling) [reply extension/jaegerremotesampling/internal/http.go:59]: I need it during the shutdown as well.

File: extension/jaegerremotesampling/internal/clientconfigmanager.go
Comment (jpkrohling) [reply extension/jaegerremotesampling/internal/clientconfigmanager.go:25]: This will still grow, so, I'll leave a comment about what it'll do in the final solution.

File: extension/jaegerremotesampling/internal/http.go
Comment (jpkrohling) [reply extension/jaegerremotesampling/internal/http.go:59]: True, but I had two things in mind:

1. I can use the mux in the tests, ensuring that the HTTP handler that I built has the right endpoints wired in. At the `Start`, I would have to provide a way to replace the HTTP server with the testing one.
2. It feels like returning a ready-to-use HTTP handler here is the right thing to do, instead of relying on the Start to complete the data structure *and* start it.

File: extension/jaegerremotesampling/internal/http.go
Comment (jpkrohling) [reply extension/jaegerremotesampling/internal/http.go:46]: At this point, we could return, as the caller of this does expect a component.Component. The tests are currently benefiting from this being a concrete type. How about we do a refactoring session once we see the whole solution ready? I think this will make sense in the next steps, but I don't want to optimize it just now.

File: extension/jaegerremotesampling/internal/http.go
Comment (jpkrohling) [reply extension/jaegerremotesampling/internal/http.go:91]: Great catch! This is indeed a deadlock, but because I wasn't ToListener on the server settings, it was just trying to bind port 80 and failing right away. So, the WG was done.

File: extension/jaegerremotesampling/internal/http_test.go
Comment (jpkrohling) [reply extension/jaegerremotesampling/internal/http_test.go:58]: I don't quite understand why they are new instances. They are both pointers to the same instance, aren't they? I didn't catch the deadlock because of a different bug (explained in an earlier comment).

File: extension/jaegerremotesampling/internal/http_test.go
Comment (bogdandrutu) [inline extension/jaegerremotesampling/internal/http_test.go:61]: Still no reason to wrap every assert in `t.Run`. This is against all the other tests, so please keep consistency :)

**Review by bogdandrutu — APPROVED** [review]

File: extension/jaegerremotesampling/internal/http_test.go
Comment (jpkrohling) [reply extension/jaegerremotesampling/internal/http_test.go]: I moved the assertions out of the t.Run's, but I would still like to understand the point about different instances being used in each block.

File: extension/jaegerremotesampling/internal/http_test.go
Comment (jpkrohling) [reply extension/jaegerremotesampling/internal/http_test.go:61]: Alright, moved out of the t.Run.

File: extension/jaegerremotesampling/internal/http.go
Comment (jpkrohling) [reply extension/jaegerremotesampling/internal/http.go:60]: I'm leaving this as is, but will change it if you don't agree with my points from the previous message.

**Comment by jpkrohling** [comment]
I'm enabling the auto-merge given that @bogdandrutu has approved this with only one extra comment, which was addressed in the latest version.

**Pull request merged**
